pkg/sql/type_change.go

```diff
@@ -612,14 +615,41 @@ func (t *typeChangeResumer) OnFailOrCancel(ctx context.Context, execCtx interfac
 		execCfg:              execCtx.(JobExecContext).ExecCfg(),
 	}
 
-	if err := tc.cleanupEnumValues(ctx); err != nil {
-		return err
+	if rollbackErr := func() error {
+		if err := tc.cleanupEnumValues(ctx); err != nil {
+			return err
+		}
+
+		if err := drainNamesForDescriptor(
+			ctx, tc.execCfg.Settings, tc.typeID, tc.execCfg.DB,
+			tc.execCfg.InternalExecutor, tc.execCfg.LeaseManager, tc.execCfg.Codec, nil,
+		); err != nil {
+			return err
+		}
+
+		if fn := tc.execCfg.TypeSchemaChangerTestingKnobs.RunAfterOnFailOrCancel; fn != nil {
+			return fn()
+		}
+
+		return nil
+	}(); rollbackErr != nil {
+		switch {
+		case errors.Is(rollbackErr, catalog.ErrDescriptorNotFound):
+			// If the descriptor for the ID can't be found, we assume that another
+			// job executed already and dropped the type.
+			log.Infof(
+				ctx,
+				"descriptor %d not found for type change job; assuming it was dropped, and exiting",
+				tc.typeID,
+			)
+		case !isPermanentSchemaChangeError(rollbackErr):
```

Comment:
i'm not sure i trust this function but i have not enough experience with it ;)
LGTM is LGT-schema!